src/tonations_generation.py

```diff
@@ -0,0 +1,203 @@
+            f"points_from_chord_patterns: {self.points_from_chord_patterns}\n"\
+            f"points_from_tail: {self.points_from_tail}"
+
+
+def tonation_to_scale(key, kind):
+    if kind == 'major':
+        return [note % 12 for note in [key, key+2, key+4, key+5, key+7, key+9, key+11]]
+    return [note % 12 for note in [key, key+2, key+3, key+5, key+7, key+8, key+10]]
+
+
+def tonation_to_tonic(key, kind):
+    if kind == 'major':
+        return [note % 12 for note in [key, key+4, key+7]]
+    return [note % 12 for note in [key, key+3, key+7]]
+
+
+def tonation_to_subdominant(key, kind):
+    if kind == 'major':
+        return [note % 12 for note in [key+5, key+8, key]]
+    return [note % 12 for note in [key+5, key+9, key]]
+
+
+def tonation_to_dominant(key, kind):
+    return [note % 12 for note in [key+7, key+11, key+2]]
+
+
+def get_tonations_from_sounds(sounds):
+    # at the moment, we only return one tonation for all sounds
+    tonations_and_points = []
+    for kind in ['major', 'minor']:
+        for base_key in range(12):
+            scale = tonation_to_scale(base_key, kind)
+            tonic = tonation_to_tonic(base_key, kind)
+            subdominant = tonation_to_subdominant(base_key, kind)
+            dominant = tonation_to_dominant(base_key, kind)
+            chord_types = get_aligned_chord_types(
+                sounds, tonic, subdominant, dominant)
+            # chord_types = remove_solitary_chords(chord_types)
+            chord_types = join_same_consecutive_chords(chord_types)
+            points_from_scale = get_points_from_scale_fit(sounds, scale)
+            points_from_chord_patterns = get_points_from_chord_patterns_fit(
+                chord_types)
+            points_from_tail = get_extra_points_for_tail(chord_types)
+            tonation = Tonation(base_key, 0.0, sounds[-1].timestamp, kind)
+            tonations_and_points.append(TonationWithPoints(
+                tonation, points_from_scale,
+                points_from_chord_patterns,
+                points_from_tail
+            ))
+    return tonations_and_points
+
+
+def get_aligned_chord_types(sounds, tonic, subdominant, dominant):
+    tend_to_tonic = True  # handles first-in-tonic = last-in-subdominant uncertainty
+    chord_types = []
+    for sound in sounds:
+        chord_type = sound_to_chord_type(
+            sound, tonic, subdominant, dominant, tend_to_tonic)
+        tend_to_tonic = chord_type != ChordType.SUBDOMINANT
+        chord_types.append(chord_type)
+    return chord_types
+
+
+def sound_to_chord_type(sound, tonic, subdominant, dominant, tend_to_tonic):
+    if sound.note in dominant:
+        return ChordType.DOMINANT
+    if sound.note in tonic:
+        if tend_to_tonic or sound.note not in subdominant:
+            return ChordType.TONIC
+        return ChordType.SUBDOMINANT
+    if sound.note in subdominant:
+        return ChordType.SUBDOMINANT
+    return ChordType.OTHER
+
+
+def remove_solitary_chords(chord_types):
```

Comment:
Shouldn't it take into consideration duration of chord? If our sound detection algorithm worked perfectly every chord would be solitary

Reply:
I addressed the first part in the comment above. Second part not true, as [(C, 0.0, 0.5), (E, 0.5, 1.0)] is interpreted as [Tonic, Tonic].